File: 3Sum.cpp
vector<vector<int> > threeSum(vector<int> &num) {
        vector<vector<int> > res;
        sort(num.begin(), num.end());
        
        int len = num.size();
        int i = 0;
        while (i < len - 2 && num[i] <= 0) {
            //while (num[i]==num[i+1]) i++;
            int j = i + 1;
            int k = len - 1;
            vector<int> tmp;
            while (j < k) {
                if (num[i]+num[j]+num[k]==0) {
                    tmp.push_back(num[i]);
                    tmp.push_back(num[j]);
                    tmp.push_back(num[k]);
                    res.push_back(tmp);
                    tmp.clear();
                    j++;
                    k--;
                    while(j<k && num[k]==num[k+1]) k--;
                    while(j<k && num[j]==num[j-1]) j++;
                } else if (num[i]+num[j]+num[k]>0) 
                    k--;
                else 
                    j++;
            }
            i++;
            while (i>0 && num[i]==num[i-1]) i++;
        }
        return res;
    }

File: 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        i = 0
        while i < len(nums)-2:
            print i
            l = self.twoSum(nums[i+1:], nums[i]*(-1))
            for ele in l:
                res.append([nums[i]]+ele)
            i+=1
            while i<len(nums)-2 and nums[i-1]==nums[i]:
                i+=1
        return res
            
    def twoSum(self, nums, target):
        res = []
        i,j = 0,len(nums)-1
        while i<j:
            if nums[i]+nums[j]==target:
                res.append([nums[i],nums[j]])
                i+=1
                while i<len(nums)-2 and nums[i-1]==nums[i]:
                    i+=1
                j-=1
                while j>1 and nums[j]==nums[j+1]:
                    j-=1
            elif nums[i]+nums[j]<target:
                i+=1
                while i<len(nums)-2 and nums[i-1]==nums[i]:
                    i+=1
            else:
                j-=1
                while j>1 and nums[j]==nums[j+1]:
                    j-=1
        return res
